lib/logstash/plugin_mixins/ecs_compatibility_support/target_check.rb

```diff
@@ -39,14 +34,63 @@ def target_set?
           false # target isn't set
         end
 
+        ##
+        # Check whether a codec is present and specifies a target.
+        #
+        # @return [nil] if codec not present or codec does not support target
+        # @return [true,false]
+        def codec_target_set?
+          return nil unless respond_to?(:codec)
+          return nil unless codec.respond_to?(:target)
+          !codec.target.nil?
+        end
+
         module RegisterDecorator
 
+          ECS_TARGET_NOT_SET_MESSAGE = ("ECS compatibility is enabled but `target` option was not specified. " +
+            "This may cause fields to be set at the top-level of the event where they are likely to clash with the Elastic Common Schema. " +
+            "It is recommended to set the `target` option to avoid potential schema conflicts " +
+            "(if your data is ECS compliant or non-conflicting, feel free to ignore this message)").freeze
+
+          ECS_TARGET_NOT_SET_PREFER_CODEC_MESSAGE = ("ECS compatibility is enabled but `target` option was not specified for this plugin or its codec." +
+            "This may cause fields to be set at the top-level of the event where they are likely to clash with the Elastic Common Schema. " +
+            "When a plugin and its codec both provide a `target` option, it is recommended to set the `target` option on the codec to avoid potential schema conflicts " +
+            "(if your data is ECS compliant or non-conflicting, feel free to ignore this message)."
+          ).freeze
+
+          MULTIPLE_TARGETS_SET_MESSAGE = ("This plugin and its codec are both configured with a `target` option, which can lead to surprising results. " +
+            "In general, it is recommended to ONLY set the codec's `target`.").freeze
+
+          PREFER_CODEC_TARGET_MESSAGE_TEMPLATE = ("Both this plugin and its codec provide a `target` option, but the codec's `target` was left unspecified. "+
+            "In general, it is recommended to only set the codec's `target`. "+
+            "To do so, remove the `target` directive from this plugin and instead define the codec with `codec => %s { target => %p }`").freeze
+
           ##
           # Logs an info message when ecs_compatibility is enabled but plugin has no `target` configuration specified.
           # @override
           def register
             super.tap do
-              logger.info(TARGET_NOT_SET_MESSAGE) if target_set? == false
+              plugin_targeted = target_set?
+              codec_targeted = codec_target_set?
+
+              if plugin_targeted
+                if codec_targeted
+                  # both is not good. prefer codec.
+                  logger.warn(MULTIPLE_TARGETS_SET_MESSAGE)
+                elsif codec_targeted == false
+                  # prefer codec's target option
+                  logger.warn(PREFER_CODEC_TARGET_MESSAGE_TEMPLATE % [codec.config_name, target])
```

Comment:
what is annoying here is that we're to see 2 warnings (few lines apart) e.g. 
```
[2021-07-19T15:45:59,339][INFO ][logstash.codecs.json     ] ECS compatibility is enabled but `target` option was not specified. This may cause fields to be set at the top-level of the event where they are likely to clash with the Elastic Common Schema. It is recommended to set the `target` option to avoid potential schema conflicts (if your data is ECS compliant or non-conflicting, feel free to ignore this message)
[2021-07-19T15:45:59,738][INFO ][logstash.javapipeline    ][main] Starting pipeline {:pipeline_id=>"main", "pipeline.workers"=>16, "pipeline.batch.size"=>10, "pipeline.batch.delay"=>50, "pipeline.max_inflight"=>160, "pipeline.sources"=>["config string"], :thread=>"#<Thread:0x6ee92a3d run>"}
Warning: Nashorn engine is planned to be removed from a future JDK release
[2021-07-19T15:46:00,852][INFO ][logstash.javapipeline    ][main] Pipeline Java execution initialization time {"seconds"=>1.09}
[2021-07-19T15:46:01,407][INFO ][logstash.inputs.http_poller][main] ECS compatibility is enabled but `target` option was not specified for this plugin or its codec.This may cause fields to be set at the top-level of the event where they are likely to clash with the Elastic Common Schema. When a plugin and its codec both provide a `target` option, it is recommended to set the `target` option on the codec to avoid potential schema conflicts (if your data is ECS compliant or non-conflicting, feel free to ignore this message).
[2021-07-19T15:46:01,417][INFO ][logstash.javapipeline    ][main] Pipeline started {"pipeline.id"=>"main"}

```
the added message here shares more information but given there's no easy way to disable the codec's warning atm I am not sure if we "really" want this.

as a side note codecs can't provide a lot of "useful" context in their logging - having the log line add an id is possible but it's the codecs id so users will need to do some juggling to figure out where to put an id to figure out where the warning is coming from (esp. if a codec is the default option).

Reply:
just to be clear - no blocker on my end shipping the feature just wanted to raise awareness about our current behavior.